src/storage/compression/bitpacking.cpp

```diff
@@ -0,0 +1,599 @@
+		                                                          offset_in_compression_group);
+
+		// TODO Optimization: avoid use of decompression buffer if we can compress straight to result vector
+		// Calculate start of compression algorithm group
+		data_ptr_t current_position_ptr =
+		    scan_state.current_width_group_ptr + scan_state.position_in_group * scan_state.current_width / 8;
+		data_ptr_t decompression_group_start_pointer =
+		    current_position_ptr - offset_in_compression_group * scan_state.current_width / 8;
+
+		// Decompress compression algorithm to buffer
+		scan_state.decompress_function((data_ptr_t)scan_state.decompression_buffer, decompression_group_start_pointer,
+		                               scan_state.current_width);
+
+		// Copy decompressed result to vector
+		T *current_result_ptr = result_data + result_offset + scanned;
+
+		if (std::is_same<T, PRE_CAST_TYPE>::value) {
+			memcpy(current_result_ptr, scan_state.decompression_buffer + offset_in_compression_group,
+			       to_scan * sizeof(T));
+		} else {
+			for (idx_t i = 0; i < to_scan; i++) {
+				current_result_ptr[i] =
+				    (T) * ((PRE_CAST_TYPE *)(scan_state.decompression_buffer + offset_in_compression_group) + i);
+			}
+		}
+
+		scanned += to_scan;
+		scan_state.position_in_group += to_scan;
+	}
+}
+
+template <class T, class PRE_CAST_TYPE>
+void BitpackingScan(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result) {
+	BitpackingScanPartial<T, PRE_CAST_TYPE>(segment, state, scan_count, result, 0);
+}
+
+//===--------------------------------------------------------------------===//
+// Fetch
+//===--------------------------------------------------------------------===//
+template <class T, class PRE_CAST_TYPE>
+void BitpackingFetchRow(ColumnSegment &segment, ColumnFetchState &state, row_t row_id, Vector &result,
+                        idx_t result_idx) {
+	BitpackingScanState<T, PRE_CAST_TYPE> scan_state(segment);
+	scan_state.Skip(segment, row_id);
+	auto result_data = FlatVector::GetData<T>(result);
+	T *current_result_ptr = result_data + result_idx;
+
+	// Todo clean up, is reused in partialscan
+	idx_t offset_in_compression_group =
+	    scan_state.position_in_group % BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE;
+
+	data_ptr_t decompression_group_start_pointer =
+	    scan_state.current_width_group_ptr +
+	    (scan_state.position_in_group - offset_in_compression_group) * scan_state.current_width / 8;
+
+	scan_state.decompress_function((data_ptr_t)scan_state.decompression_buffer, decompression_group_start_pointer,
+	                               scan_state.current_width);
+
+	if (std::is_same<T, PRE_CAST_TYPE>::value) {
+		*current_result_ptr = *(T *)(scan_state.decompression_buffer + offset_in_compression_group);
+	} else {
+		*current_result_ptr = (T) * (PRE_CAST_TYPE *)(scan_state.decompression_buffer + offset_in_compression_group);
```

Comment:
This line is uncovered, could you add a test that covers it?

Reply:
sure thing!